### ISSUE TITLE
Improved `NotEqual` operator

### ISSUE THREAD

**Issue by etiennedi** [weaviate/weaviate#3319]
The current `NotEqual` operator has two issues:

1. It's slow and doesn't scale as it performs a full-scan on the inverted index
2. It's incorrect on tokenized text, e.g. `I like Wikipedia` is tokenized into `I`, `like`, `wikipedia`. All point to the same document. Therefore searching for `NotEqual "Wikipedia"` returns the document.

## Proposed improvements

As discussed with @aliszka, we could use bitmaps. Pre-populate a bitmap with all possible doc-ids from 0 to max_doc_id. Then subtract the matches. This will lead to some false positives (e.g. IDs that don't exist anymore). This is not a problem for filtered vector search (where we only build an allow list), but it is a problem for pure `where` filters, where we could end up with too few matches. As discussed with @aliszka this could be mitigated by iterating over ids rather than having a fixed cut off in such a scenario.

## Acceptance Criteria

- Open a PR following the implementation suggested above - any alternative implementation needs to be approved by us first.
- The patch needs to be objectively faster than the old solution. This means some performance tests should be included with results, proving that the change is truly an improvement. 
- There should be unit and acceptance tests demonstrating that queries with tokenized text work correctly.
- The solution must work with pure-filtered and vector-filtered search.
    - See [here](https://github.com/weaviate/weaviate/issues/3319#issuecomment-1787065036) for more context
- The preview docker image of your branch should pass the [chaos pipeline](https://github.com/weaviate/weaviate-chaos-engineering) (we will do this internally as a final validation step).

**Comment by princejha95** [comment]
@etiennedi Can i pick this up ?

**Comment by etiennedi** [comment]
@princejha95 yes, gladly. Do you have a good understanding of how the current version works and how bitmaps work in Weaviate? Does the "Proposed Improvements" section above give you enough context?

**Comment by princejha95** [comment]
@etiennedi I have understood the problem and the suggestion given in "Proposed Improvements" section. Regarding "how the current version works" and "how bitmap works in Weaviate", i will develop my understanding by going through the codebase. But if you have any documentation available in tihs regard, please do share.

Thanks.
Prince

**Comment by etiennedi** [comment]
Thanks for elaborating. It sounds like you have a good understanding of the problem and we can help you on the parts that aren't clear yet. We'll come back to those points shortly.

**Comment by etiennedi** [comment]
In the meantime, please also take a look at [this branch](https://github.com/weaviate/weaviate/commits/not-equal-bitmap) where @aliszka and I already got started on this. You can reuse this for your contribution. I think it's already most of the way there, we just need to deal with some of the edge cases and add plenty more tests.

**Comment by etiennedi** [comment]
> I have not completely understood This is not a problem for filtered vector search (where we only build an allow list), but it is a problem for pure where filters, where we could end up with too few matches

Let me explain in more detail:

The approach outlined above and also on branch `not-equal-bitmap` is that we take the highest possible doc id and populate it with all possible values between 0 and `max_doc_id`. For example, if the max doc is 99, the bit map would look something like

| bit | 0 | 1 | 2 | 3 | 4 | ... | 95 | 96 | 97 | 98 | 99 |
| --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- |
| value | 1 | 1 | 1 | 1 | 1 | ... | 1 | 1 | 1 | 1 | 1 | 

It contains 100 `1`s and 0 `0`s

Then we substract all the matches for an Equal filter. For the sake of simplicitly, let's say the Equal filter matched every even doc id. After substracting all even doc ids our bitmap looks something like this:

| bit | 0 | 1 | 2 | 3 | 4 | ... | 95 | 96 | 97 | 98 | 99 |
| --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- |
| value | 0 | 1 | 0 | 1 | 0 | ... | 1 | 0 | 1 | 0 | 1 |

The bitmap now contains 50 `1`s and 50 `0`s.

However, there is one other problem. We just assumed in step that every ID between 0 and 99 exists. This is not guaranteed to be true. For example, a user could have deleted all ids between 0-9. We have no way of knowing this. So in our bitmaps we actually have a few false positives. The ids 1,3,5,7,9 refer to objects that no longer exists.

Above we said this is not a problem for a filtered vector seach, but it is a problem for a pure filter search? Why is that? – Let me explain.

In the case of a filtered vector search we only ever use the bitmap as an allow list. In other words the vector search alogrithm would ask the bitmap "am I allowed to use id 42?" the bitmap would say "no", "am I allowed to use id 43?", the bitmap would say "yes". The fact that there are false positives on the bitmap doesn't matter because no one will ever ask for them.

When we do a pure filtered search (without a vector search), this is different: Now the bitmap becomes the source of objects to return. For example if the user sets a limit of 10, we would take the first 10 items on the bitmap. In our case, they would be: 1,3,5,7,9,11,13,15,17,19. We identified previously that the first 5 of those no longer exist. So the user would only see 5 results (11,13,15,17,19) because 1,3,5,7,9 don't exist. This would be a problem because the user asked for 10 results.

To fix this, we cannot apply the limit this early. Instead we need to change the algorithm to be a simple cursor on the bitmap with a dynamic exit condition. It should not stop after retrieving 10 ids, but rather it should stop after it retrieved 10 ids that could resolve to an actual object. In our example above that would require retrieving 15 ids. 

Hope this helps!

**Comment by princejha95** [comment]
Thanks a lot @etiennedi for the explanation. This logic is now crystal clear to me.

**Comment by etiennedi** [comment]
Hi @princejha95. I was curious if you were able to make any progress on this or if you need any more help. Thanks.

**Comment by parkerduckworth** [comment]
Some additional depth about the root cause of the tokenized text issue with the current implementation:

Given the phrase `"I like Wikipedia"` from document `X` and tokenizing it -> `["I", "like", "Wikipedia"]`, each token is indexed separately, and each inverted index entry contains `X` in the posting list:

```
"I"         -> [..., X, ...]
"like"      -> [..., X, ...]
"Wikipedia" -> [..., X, ...]
```

The current `NotEqual` implementation will consider each inverted index entry separately. Because all three tokens in this example map to a posting list containing document `X`, the current algorithm will place `X` back into the results list once all three inverted index entries have been analyzed.

In other words, let's say that in our example, documents containing the token `"I"` are excluded from the result set, so `X` is rightly left out. But then, when analyzing documents that don't contain `"like:"`, `"I"` is not equal to `"like"`, so `X` is added back to the result set, and so the process continues.

The proposed solution is expected to produce a result set such that `X` is excluded irrespective of the total set of input tokens.

**Comment by etiennedi** [comment]
Hi @princejha95, if you are still working on this, please let us know before the end of the week. Otherwise, we assume that this issue is up for grabs for someone else. Thanks.

**Comment by SingL3** [comment]
Any update here?

**Comment by parkerduckworth** [comment]
@SingL3 actively under development. Will be included in the 1.24 release targeted for mid-Feb

**Comment by parkerduckworth** [comment]
Fixed by #4122

**Comment by GameSetAndMatch** [comment]
Will this include this requested feature: https://github.com/weaviate/weaviate/issues/3683 ?

**Comment by parkerduckworth** [comment]
@GameSetAndMatch the groundwork for implementing such a feature has been implemented with #4122, but the exact behavior expressed as desired in #3683 has not.

Likely we will apply a similar approach as we did with `Equal`/`NotEqual` to `Like`, by introducing a `NotLike` operator, rather than an independent `Not` operator, but still need more discussion surrounding the specifics of the implementation.
